File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
            if head == None or head.next == None:
                return head
            else:
                # get the reverse sublist, withthout head attached
                reversedListHead = self.reverseList(head.next)
                # the current tail is head.next
                # make a variable for the tail
                tail = head.next
                # point tail to the former head node as the next in line
                tail.next = head 
                # clear head's next pointer, since it is now the new tail 
                head.next = None
                return reversedListHead

File: Data Structures & Algorithms/reorder-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
   public:
    void reorderList(ListNode* head) {
        ListNode* l1 = head;
        ListNode* l2 = head->next;
        ListNode* prev = nullptr;
        ListNode* current = head;
        ListNode* slow = head;
        ListNode* fast = head->next;

        // find the middle node in the list
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        l2 = slow->next;
        slow->next = nullptr;

        // reverse the 2nd list
        current = l2;
        while (current) {
            ListNode* temp = current->next;
            current->next = prev;
            prev = current;
            current = temp;
        }

        while (l1 && prev) {
            ListNode * temp1 = l1->next;
            ListNode * temp2 = prev->next;
            l1->next = prev;
            prev->next = temp1;
            l1 = temp1;
            prev = temp2;
        }
    }
};
